shared/src/balancer/event_handler.rs

```diff
@@ -120,22 +124,22 @@ impl PoolSpecialization {
     }
 }
 
-impl BalancerPools {
-    fn try_upgrade(&mut self) {
+impl PoolRegistry {
+    fn try_upgrade(&mut self) -> Result<()> {
         for (pool_id, pool_builder) in self.pending_pools.clone() {
-            if let Some(weighted_pool) = pool_builder.into_pool() {
-                // When upgradable, delete pending pool and add to valid pools
-                tracing::info!("Upgrading Pool Builder with id {:?}", pool_id);
-                self.pools.insert(pool_id, weighted_pool.clone());
-                self.pending_pools.remove(&pool_id);
-                for token in weighted_pool.tokens {
-                    self.pools_by_token
-                        .entry(token)
-                        .or_default()
-                        .insert(pool_id);
-                }
+            let weighted_pool = pool_builder.into_pool()?;
```

Comment:
won't this return if a single pool fails to be turned into a weighted pool (also is the variable name outdated now?) ? Maybe add a unit test for this if not too much hassle? 

I guess this is fine, because we expect events to be part of the same transaction, but I remember that we used to be able to handle the case even if they were split across multiple pages.

Reply:
Now that you mention it I do slight prefer returning the option and logging an error when upgrade fails.